runner/run.go

```diff
@@ -1,80 +1,209 @@
-type Run struct {
-	// Information about the Taskcluster deployment where this
-	// worker is runing
-	RootURL string
-
-	// Credentials for the worker, and their expiration time.  Shortly before
-	// this expiration, worker-runner will try to gracefully stop the worker
-	Credentials       taskcluster.Credentials
-	CredentialsExpire time.Time
-
-	// Information about this worker
-	WorkerPoolID string
-	WorkerGroup  string
-	WorkerID     string
-
-	// metadata from the provider (useful to display to the user for
-	// debugging).  Workers should not *require* any data to exist
-	// in this map, and where possible should just pass it along as-is
-	// in worker config as helpful debugging metadata for the user.
-	ProviderMetadata map[string]string
-
-	// the accumulated WorkerConfig for this run, including files to create
-	WorkerConfig *cfg.WorkerConfig
-	Files        []files.File
-
-	// the protocol (set in SetProtocol)
-	proto *protocol.Protocol
-
-	// a timer to handle sending a graceful-termination request before
-	// the credentials expire
-	credsExpireTimer *time.Timer
-}
+// Check that the provider filled the run fields it was expected to.
+func checkProviderResults(run *run.Run) error {
+	if run.RootURL == "" {
+		return fmt.Errorf("provider did not set RootURL")
+	}
 
-func (r *Run) SetProtocol(proto *protocol.Protocol) {
-	r.proto = proto
-}
+	if run.Credentials.ClientID == "" {
+		return fmt.Errorf("provider did not set Credentials.ClientID")
+	}
+
+	if run.WorkerPoolID == "" {
+		return fmt.Errorf("provider did not set WorkerPoolID")
+	}
+
+	if run.WorkerGroup == "" {
+		return fmt.Errorf("provider did not set WorkerGroup")
+	}
+
+	if run.WorkerID == "" {
+		return fmt.Errorf("provider did not set WorkerID")
+	}
 
-func (r *Run) WorkerStarted() error {
-	// gracefully terminate the worker when the credentials expire, if they expire
-	if r.CredentialsExpire.IsZero() {
-		return nil
-	}
-
-	untilExpire := time.Until(r.CredentialsExpire)
-	r.credsExpireTimer = time.AfterFunc(untilExpire-30*time.Second, func() {
-		if r.proto != nil && r.proto.Capable("graceful-termination") {
-			log.Println("Taskcluster Credentials are expiring in 30s; stopping worker")
-			r.proto.Send(protocol.Message{
-				Type: "graceful-termination",
-				Properties: map[string]interface{}{
-					// credentials are expiring, so no time to shut down..
-					"finish-tasks": false,
-				},
-			})
-		}
-	})
 	return nil
 }
 
-func (r *Run) WorkerFinished() error {
-	if r.credsExpireTimer != nil {
-		r.credsExpireTimer.Stop()
-		r.credsExpireTimer = nil
+// Run the worker.  This embodies the execution of the start-worker command.
+func Run(configFile string) (run run.Run, err error) {
+	// load configuration
+
+	log.Printf("Loading taskcluster-worker-runner configuration from %s", configFile)
+	runnercfg, err := cfg.LoadRunnerConfig(configFile)
+	if err != nil {
+		err = fmt.Errorf("Error lading runner config file %s: %s", configFile, err)
```

Comment:
nit: "loading"

Reply:
Yep, fixed in the PR for that commit, thanks :)